### PR TITLE
Add conversion for tosa.depthwise_conv2d 

### PR DESCRIPTION
**Pull request by dajuro**
Adds a conversion for `tosa.depthwise_conv2d`, the corresponding unittest and regression test. Also adds utility functions to calculate the index to a flat array for a given shape and indices (ravel index). This is needed in the `depthwise_conv2d` implementation.

### CODE REVIEW

File: include/emitc/emitc_types.h [hunk 1]
```diff
@@ -396,4 +396,46 @@ struct concat<Dim, T, Tensor4D<T, D0, D1, D2, D3>...> {
...
+
+  std::array<size_t, sizeof...(Shape)> s = strides<Shape...>();
+
+  size_t result = 0;
+  for (size_t i = 0; i < indicesArray.size(); ++i) {
+    result += indicesArray[i] * s[i];
+  }
+
+  return result;
+}
```
Comment (simon-camp) [inline include/emitc/emitc_types.h:439]: Are these copied 1:1 from the Tensor class? If so we can remove the implementation from the class. Also  add `unravel_index` for completeness.

File: include/emitc/emitc_tosa.h [hunk 2]
```diff
@@ -178,6 +178,91 @@ Dest conv2d(Src input, Weights weights, Tensor1D<int64_t, 4> padding,
...
+  assert(stride[1] > 0);
+
+  assert(dilation[0] == 1);
+  assert(dilation[1] == 1);
+
+  const int N = input.dim(0);
+  const int H_IN = input.dim(1);
+  const int W_IN = input.dim(2);
+
+  const int feature_group_count = input.dim(3);
```
Comment (simon-camp) [inline include/emitc/emitc_tosa.h:205]: Rename this to C_IN

File: include/emitc/emitc_tosa.h [hunk 3]
```diff
@@ -178,6 +178,91 @@ Dest conv2d(Src input, Weights weights, Tensor1D<int64_t, 4> padding,
...
+
+  Dest output;
+
+  const int C_OUT = output.dim(3);
+  assert(C_OUT % feature_group_count == 0);
+  const int G_OUT = C_OUT / feature_group_count;
```
Comment (simon-camp) [inline include/emitc/emitc_tosa.h:211]: MULTIPLIER

File: include/emitc/emitc_tosa.h [hunk 4]
```diff
@@ -178,6 +178,91 @@ Dest conv2d(Src input, Weights weights, Tensor1D<int64_t, 4> padding,
...
+  const int H_PAD = pt + H_IN + pb;
+  const int W_PAD = pl + W_IN + pr;
+
+  // Convolution
+  for (int n = 0; n < N; ++n) {
+    for (int h_pad = 0; h_pad < H_PAD - K_H + 1; h_pad += S_H) {
+      for (int w_pad = 0; w_pad < W_PAD - K_W + 1; w_pad += S_W) {
+        for (int kh = 0; kh < K_H; ++kh) {
+          for (int kw = 0; kw < K_W; ++kw) {
+            for (int g = 0; g < feature_group_count; ++g) {
```
Comment (simon-camp) [inline include/emitc/emitc_tosa.h:233]: c_in

File: include/emitc/emitc_tosa.h [hunk 5]
```diff
@@ -178,6 +178,91 @@ Dest conv2d(Src input, Weights weights, Tensor1D<int64_t, 4> padding,
...
+              for (int g_out = 0; g_out < G_OUT; ++g_out) {
```
Comment (simon-camp) [inline include/emitc/emitc_tosa.h:234]: m

File: include/emitc/emitc_tosa.h [hunk 6]
```diff
@@ -178,6 +178,91 @@ Dest conv2d(Src input, Weights weights, Tensor1D<int64_t, 4> padding,
...
+                const int w_in = w_pad - pl + kw;
+
+                if (h_in < 0 || h_in >= H_IN || w_in < 0 || w_in >= W_IN)
+                  continue;
+
+                // for depthwise convolution we interpret weights as a tensor
+                // with shape [filter_height, filter_width, 1, in_channels *
+                // channel_multiplier]. So we need to calculate the index
+                // using these dimensions.
+                const int weights_index =
```
Comment (OliverScherf) [inline include/emitc/emitc_tosa.h:248]:
```suggestion
                const size_t weights_index =
```

File: include/emitc/emitc_tosa.h [hunk 7]
```diff
@@ -178,6 +178,91 @@ Dest conv2d(Src input, Weights weights, Tensor1D<int64_t, 4> padding,
   return output;
 }
 
+// DepthwiseConv2DOp
+template <typename Dest, typename Src, typename Weights>
+Dest depthwise_conv2d(Src input, Weights weights, Tensor1D<int64_t, 4> padding,
+                      Tensor1D<int64_t, 2> stride,
+                      Tensor1D<int64_t, 2> dilation) {
+  // Input is [N,IH,IW,IC], weights
```
Comment (OliverScherf) [inline include/emitc/emitc_tosa.h:186]: Why are you not using the same notation as the constants in the function itself?

`IH` -> `H_IN` e.g.

File: include/emitc/emitc_tosa.h [hunk 8]
```diff
@@ -178,6 +178,91 @@ Dest conv2d(Src input, Weights weights, Tensor1D<int64_t, 4> padding,
...
+  // are [KH,KW,CIN,M] and output is [N,H,W,C*M].
```
Comment (simon-camp) [inline include/emitc/emitc_tosa.h:187]: Optional: You could add static_asserts that the dimensions match
```c
Src::dim(0) == Dest::dim(0);
Src::dim(3) == Weights::dim(2);
...
```

File: include/emitc/emitc_tosa.h [hunk 9]
```diff
@@ -178,6 +178,91 @@ Dest conv2d(Src input, Weights weights, Tensor1D<int64_t, 4> padding,
...
+  const int N = input.dim(0);
+  const int H_IN = input.dim(1);
+  const int W_IN = input.dim(2);
+
+  const int feature_group_count = input.dim(3);
+
+  Dest output;
+
+  const int C_OUT = output.dim(3);
+  assert(C_OUT % feature_group_count == 0);
```
Comment (simon-camp) [inline include/emitc/emitc_tosa.h:210]: This can be a static_assert at the top (call the dim method on the Type).

**Review by simon-camp — CHANGES_REQUESTED** [review]
Looks good overall, mostly renaming a few variables :)

File: include/emitc/emitc_types.h (hunk 1 above)
Comment (dajuro) [reply include/emitc/emitc_types.h:439]: Mostly, the parameters of the functions are different (`std::array` vs variadic arguments). That's why I didn't remove it from the class. Can we forward `std::array` to variadic?

File: include/emitc/emitc_tosa.h (hunk 8 above)
Comment (dajuro) [reply include/emitc/emitc_tosa.h:187]: That's to obvious :P

**Review by OliverScherf — CHANGES_REQUESTED** [review]
Some NIT in addition to @simon-camp's comments.

File: include/emitc/emitc_types.h (hunk 1 above)
Comment (simon-camp) [reply include/emitc/emitc_types.h:439]: I'm pretty sure there is some meta programming voodoo you can do, or you could add an overload, or you keep it as is and we come back to this later.

File: lib/Dialect/EmitC/Conversion/TosaToEmitC.cpp [hunk 10]
```diff
@@ -108,16 +108,17 @@ class ConstOpConversion : public OpRewritePattern<tosa::ConstOp> {
...
-class Conv2DOpConversion : public OpConversionPattern<tosa::Conv2DOp> {
-  using OpConversionPattern<tosa::Conv2DOp>::OpConversionPattern;
+template <typename SrcOp>
+class Conv2DOpConversion : public OpConversionPattern<SrcOp> {
+  using OpConversionPattern<SrcOp>::OpConversionPattern;
 
 public:
-  Conv2DOpConversion(MLIRContext *ctx)
-      : OpConversionPattern<tosa::Conv2DOp>(ctx) {}
+  Conv2DOpConversion(MLIRContext *ctx, StringRef funcName)
```
Comment (marbre) [inline lib/Dialect/EmitC/Conversion/TosaToEmitC.cpp:116]:
```suggestion
  ConvOpConversion(MLIRContext *ctx, StringRef funcName)
```

If it's not only for a 2D conv anymore, this should be renamed.

File: lib/Dialect/EmitC/Conversion/TosaToEmitC.cpp [hunk 11]
```diff
@@ -108,16 +108,17 @@ class ConstOpConversion : public OpRewritePattern<tosa::ConstOp> {
...
+      : OpConversionPattern<SrcOp>(ctx), funcName(funcName) {}
 
 private:
   LogicalResult
-  matchAndRewrite(tosa::Conv2DOp conv2dOp, ArrayRef<Value> operands,
+  matchAndRewrite(SrcOp conv2dOp, ArrayRef<Value> operands,
```
Comment (marbre) [inline lib/Dialect/EmitC/Conversion/TosaToEmitC.cpp:121]:
```suggestion
  matchAndRewrite(SrcOp convOp, ArrayRef<Value> operands,
```

File: lib/Dialect/EmitC/Conversion/TosaToEmitC.cpp [hunk 12]
```diff
@@ -108,16 +108,17 @@ class ConstOpConversion : public OpRewritePattern<tosa::ConstOp> {
...
                   ConversionPatternRewriter &rewriter) const override {
     // fail if quantization is requested
     if (conv2dOp.quantization_info().hasValue()) {
```
Comment (marbre) [inline lib/Dialect/EmitC/Conversion/TosaToEmitC.cpp:124]:
```suggestion
    if (convOp.quantization_info().hasValue()) {
```

Comment (marbre) [reply lib/Dialect/EmitC/Conversion/TosaToEmitC.cpp:124]: This requires some further renaming at other places.

**Review by marbre — CHANGES_REQUESTED** [review]

File: include/emitc/emitc_tosa.h (hunk 7 above)
Comment (dajuro) [reply include/emitc/emitc_tosa.h:186]: Yep, I agree: that's confusing.

File: lib/Dialect/EmitC/Conversion/TosaToEmitC.cpp (hunk 12 above)
Comment (simon-camp) [reply lib/Dialect/EmitC/Conversion/TosaToEmitC.cpp:124]: Then also change the template parameter name to ConvOp.

File: lib/Dialect/EmitC/Conversion/TosaToEmitC.cpp (hunk 10 above)
Comment (dajuro) [reply lib/Dialect/EmitC/Conversion/TosaToEmitC.cpp:116]: `depthwise_conv2d` is still a `conv2d` :P

Comment (marbre) [reply lib/Dialect/EmitC/Conversion/TosaToEmitC.cpp:116]: Yes, but the name maps to closely to `Conv2DOp` which is somehow misleading.

Comment (dajuro) [reply lib/Dialect/EmitC/Conversion/TosaToEmitC.cpp]: I renamed it to `GenericConvOpConversion`

File: include/emitc/emitc_tosa.h [hunk 13]
```diff
@@ -178,6 +178,91 @@ Dest conv2d(Src input, Weights weights, Tensor1D<int64_t, 4> padding,
...
+
+  const int N = input.dim(0);
+  const int H_IN = input.dim(1);
+  const int W_IN = input.dim(2);
+  const int C_IN = input.dim(3);
+
+  Dest output;
+
+  const int CxM = output.dim(3);
+  const int M = CxM / C_IN;
```
Comment (simon-camp) [inline include/emitc/emitc_tosa.h:212]: `M = Weights::dim(3)`?

**Review by simon-camp — CHANGES_REQUESTED** [review]

File: include/emitc/emitc_tosa.h [hunk 14]
```diff
@@ -178,6 +178,91 @@ Dest conv2d(Src input, Weights weights, Tensor1D<int64_t, 4> padding,
...
+  // are [K_H,K_W,C_IN,M] and output is [N,H,W,C_IN*M].
+  static_assert(is_tensor_of_dim<4, Src>::value,
+                "Expected 4 dimensional input");
+  static_assert(is_tensor_of_dim<4, Dest>::value,
+                "Expected 4 dimensional output");
+  static_assert(is_tensor_of_dim<4, Weights>::value,
+                "Expected 4 dimensional weights");
+  static_assert(Src::dim(0) == Dest::dim(0), "Batch sizes must be equal");
+  static_assert(Dest::dim(3) % Src::dim(3) == 0,
+                "Output channels need to be a multiple of input channels");
```
[the remark this answers is not in the document]
Comment (dajuro) [reply include/emitc/emitc_tosa.h:196]: I agree to the first one. The second one doesn't work ;)

File: include/emitc/emitc_tosa.h (hunk 13 above)
Comment (dajuro) [reply include/emitc/emitc_tosa.h:212]: Okay, yeah that's much better :D

File: include/emitc/emitc_tosa.h (hunk 14 above)
[the remark this answers is not in the document]
Comment (dajuro) [reply include/emitc/emitc_tosa.h:196]: Ok nevermind... the unittest was buggy

**Review by simon-camp — APPROVED** [review]

**Review by OliverScherf — APPROVED** [review]

**Review by marbre — APPROVED** [review]

**Pull request merged**